airframe/src/test/scala/wvlet/airframe/CloseableShutdownHookTest.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package wvlet.airframe
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import wvlet.airspec.AirSpec
+
+/**
+  *
+  */
+class CloseableShutdownHookTest extends AirSpec {
+  scalaJsSupport
+
+  class A extends AutoCloseable {
+    val closeIsCalled = new AtomicBoolean(false)
+
+    override def close(): Unit = {
+      closeIsCalled.set(true)
+    }
+  }
+
+  def `support closeable`: Unit = {
+    val a = new A
+    val d = newSilentDesign.bind[A].toInstance(a)
+    d.build[A] { a =>
+    }
+
+    a.closeIsCalled.get() shouldBe true
+  }
+
+  trait A1 {
+    val onShutdownIsCalled = new AtomicBoolean(false)
+    val a = bind[A].onShutdown { x =>
+      onShutdownIsCalled.set(true)
+    }
```

Comment:
How does the shutdown process work if a singleton component is bound in multiple places with `onShutdown` hook? For example:

```scala
trait A1 {
  val a = bind[A].onShutdown { ... }
}

trait A2 {
  val a = bind[A].onShutdown { ... }
}

val design = newDesign
  .bind[A].toSingleton // or toInstance(), or simply bind[A]?
```

Reply:
If A is a singleton, only the first registered shutdown hook will be used.

For example, if A1 is created first, its onShutdown hook will be used and A2's shutdown hook will be ignored.

Reply:
I usually feel difficulties to find an appropriate place to write such shutdown hooks. As in your examples, we may define duplicate shutdown hooks and its behavior can be confusing, but at the same time, we should not call shutdown hooks multiple times. 

So the purpose of this PR is for using a pre-defined resource deallocation method (close()) if exists. I think this will reduce the needs for defining onShutdown hooks.

Reply:
Yes, I agree with difficulties in defining shutdown hooks at binding-time. But a way to define shutdown hook is still necessary because making all components which need shutdown process is no reality. So I think we would be better to deprecate the binding-time shutdown hook definition and encourage to use the defining-time shutdown hook definition.

Reply:
I also think adding shutdown hooks to a design is a good idea (I sometimes want to do that). I’ll file a ticket